File: UDP/udp_client.py
#!/usr/bin/env python
import socket
import sys
import struct
import random
import time
from struct import *

port = 10022
max_data_size = 100
request_id = 1

# check to see if you have valid command line inputs
if (len(sys.argv) != 3):
     sys.stderr.write("usage: talker hostname message\n")
     exit(1)

#set the values of the two argument into variables
serverName = sys.argv[1]
portNumber = int(sys.argv[2])

#Ask user to send a message or exit
send_message = input("To send a message, press 1. To exit, press 2.")
while(send_message != 2):
    #check if its an ipv4 or ipv6 Address
    ipType = 0
    try:
        socket.inet_aton(serverName)
        ipType = 4
    except socket.error:
        ipType = 6

    if ipType != 4:
        try:
            socket.inet_pton(socket.AF_INET6, serverName)
        except socket.error:
            ipType = 0

    if(ipType == 0):
        sys.stderr.write('Not a valid ip\n')
        exit(1)

    if(ipType == 4):
        ip = socket.inet_aton(serverName)
        family = socket.AF_INET
    else:
        ip = socket.inet_pton(socket.AF_INET6, serverName)
        family = socket.AF_INET6


    # Create a datagram socket
    try:
        s = socket.socket(family, socket.SOCK_DGRAM)
    except socket.error:
        sys.stderr.write('talker: Failed to create socket')
        sys.exit()


    total_message_length = 8
    op_code = 8
    num_operands = 8
    op_1 = 16
    op_2 = 16

    opCode = eval(input("Please enter an Opcode between 0 and 6: "))
    while (opCode < 0 or opCode > 6):
        opCode = eval(input("Invalid Entry: Please enter an Opcode between 0 and 6: "))

    op1 = eval(input("Please enter operand 1: "))

    if (opCode == 6):
        total_message_length = 6
        num_operands = 1
        op2 = 0

    else:
        total_message_length = 8
        num_operands = 2
        op2 = eval(input("Please enter operand 2: "))

    #op1 = eval(input("Please enter operand 1: "))

    op_code = opCode

    op_1 = op1

    op_2 = op2

    #Pack the message into a Struct in Big Endian
    message = pack('!bbbbhh', total_message_length, request_id , op_code, num_operands, op_1, op_2)

    #Send a message and measure time it takes to send a message
    send_start = time.time()
    send = s.sendto(message, (serverName, portNumber))
    send_time = time.time()- send_start
    print ("Time to send: {}".format(send_time))

    #Check to see if you can send
    if(send <= 0):
        sys.stderr.write("Send\n")
        exit(1)

    print("Connecting to Server: {}".format(serverName))
    print ("Message being sent is: {}".format(message))

    #Make sure valid msg, recv the message and measure time it takes to recv a message
    recv_start = time.time()
    try:
        numBytes = s.recv(max_data_size)
    except socket.error:
        sys.stderr.write("Error: Recieved\n")
        sys.exit()

    recv_time = time.time()- recv_start
    print ("Time to recv: {}".format(recv_time))

    print ("Message being received is: {}".format(numBytes))

    #not sure about this
    if (numBytes == ''):
        sys.stderr.write("recv\n")

    request_id += 1
    s.close()

exit(1)

File: UDP/UDP-client.c
/*
** talker.c -- a datagram "client" demo
*/

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <errno.h>
#include <string.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <netdb.h>

#define SERVERPORT "10022"	// the port users will be connecting to

struct message_request
{
  uint8_t total_message_length;
  uint8_t request_id;
  uint8_t op_code;
  uint8_t num_operands;
  short op_1;
  short op_2;
} __attribute__((__packed__));

int main(int argc, char *argv[])
{
	int sockfd;
	struct addrinfo hints, *servinfo, *p;
	struct sockaddr_storage their_addr;
	int rv;
	int numbytes;
	struct message_request req;
	char buffer[8];

	if (argc != 3) {
		fprintf(stderr,"usage: talker hostname message\n");
		exit(1);
	}

	memset(&hints, 0, sizeof hints);
	hints.ai_family = AF_UNSPEC;
	hints.ai_socktype = SOCK_DGRAM;

	if ((rv = getaddrinfo(argv[1], SERVERPORT, &hints, &servinfo)) != 0) {
		fprintf(stderr, "getaddrinfo: %s\n", gai_strerror(rv));
		return 1;
	}

	// loop through all the results and make a socket
	for(p = servinfo; p != NULL; p = p->ai_next) {
		if ((sockfd = socket(p->ai_family, p->ai_socktype,
				p->ai_protocol)) == -1) {
			perror("talker: socket");
			continue;
		}

		break;
	}

	if (p == NULL) {
		fprintf(stderr, "talker: failed to create socket\n");
		return 2;
	}

	req.total_message_length = 8;
	req.op_code = 0;
	req.num_operands = 2;
	req.request_id = 0;
	req.op_1 = htons(120);
	req.op_2 = htons(60);

	short op1 = htons(240);
	short op2 = htons(160);



	buffer[0] = 8;
	buffer[1] = 0;
	buffer[2] = 0;
	buffer[3] = 2;
	buffer[4] = op1 & 0xFF;
	buffer[5] = op1 >> 8;
	buffer[6] = op2 & 0xFF;
	buffer[7] = op2 >> 8;


	if ((numbytes = sendto(sockfd, buffer, sizeof(buffer), 0,
			 p->ai_addr, p->ai_addrlen)) == -1) {
		perror("talker: sendto");
		exit(1);
	}

	freeaddrinfo(servinfo);
	char message[100];
	socklen_t addr_len = sizeof their_addr;
	recvfrom(sockfd, message, 100, 0, (struct sockaddr *)&their_addr, &addr_len);

	printf("New message: \"%s\"\n", message);

	printf("talker: sent %d bytes to %s\n", numbytes, argv[1]);
	close(sockfd);

	return 0;
}
